pkg/discovery/server/proxy.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+
+	"github.com/pingcap/tidb-operator/pkg/client/clientset/versioned"
+	"github.com/pingcap/tidb-operator/pkg/manager/member"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+func buildUrl(cli versioned.Interface, tcName, namespace string) (*url.URL, error) {
+	url := &url.URL{
+		Host:   fmt.Sprintf("%s-pd:2379", tcName),
+		Scheme: "http",
+	}
+	tc, err := cli.PingcapV1alpha1().TidbClusters(namespace).Get(tcName, metav1.GetOptions{})
+	if err != nil {
+		klog.Error(err)
+		return nil, err
+	}
+	if tc.IsTLSClusterEnabled() {
+		url.Scheme = "https"
+	}
+	return url, nil
+}
+
+func buildProxy(cli versioned.Interface, kubeCli kubernetes.Interface, tcName, namespace string) *httputil.ReverseProxy {
+	url, err := buildUrl(cli, tcName, namespace)
+	if err != nil {
+		klog.Fatal(err)
+	}
+	proxy := httputil.NewSingleHostReverseProxy(url)
+	if url.Scheme == "https" {
+		// load crt and key
+		certPath := fmt.Sprintf("%s/tls.crt", member.PdTlsCertPath)
+		keyPath := fmt.Sprintf("%s/tls.key", member.PdTlsCertPath)
+		cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+		if err != nil {
+			klog.Fatal(err)
+		}
+		// load ca
+		rootCAs := x509.NewCertPool()
+		caPath := fmt.Sprintf("%s/ca.crt", member.PdTlsCertPath)
+		caByte, err := ioutil.ReadFile(caPath)
+		if err != nil {
+			klog.Fatal(err)
+		}
+		rootCAs.AppendCertsFromPEM(caByte)
+		tlsConfig := &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			RootCAs:      rootCAs,
+		}
+		proxy.Transport = &http.Transport{TLSClientConfig: tlsConfig}
+	}
```

Comment:
certs are loaded once at startup, I guess we should reload TLS certificates for every client request

Reply:
this is a simple mechanism to make sure this proxy will not fail when the certs are expired
we can also have some unit tests to verify this
ref: https://github.com/etcd-io/etcd/pull/7829

Reply:
> certs are loaded once at startup, I guess we should reload TLS certificates for every client request

Not sure whether this would be quite time-consuming but is this necessary?

Reply:
this is how PD/TiDB refresh TLS certificates too
it's possible to refresh on expiration (see etcd TODO), but I don't think it's necessary. just read certificates every time
actually these small disk files are cached in buffer in most time

Reply:
make sense. updated. now the proxy would be initialed each time